File: Blind 75/03 - Sliding Window/121-best-time-to-buy-and-sell-stock.py
'''
You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
'''
from time import time
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        i = 0
        j = i + 1
        while i < len(prices) and j < len(prices):
            profit = prices[j] - prices[i]
            if profit < 0:
                i = j
                continue
            elif profit > 0 and profit > max_profit:
                max_profit = profit
            j += 1

        return max_profit

    def reference(self, prices: List[int]) -> int:
        res = 0
        l = 0
        for r in range(1, len(prices)):
            if prices[r] < prices[l]:
                l = r
            res = max(res, prices[r] - prices[l])

        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.maxProfit(case))
                else:
                    self.maxProfit(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [7, 1, 5, 3, 6, 4],
        [7, 6, 4, 3, 1]
    ]
    test.quantify(test_cases)

File: Blind 75/03 - Sliding Window/JavaScript/121-best-time-to-buy-and-sell-stock.js
const { performance } = require('perf_hooks');

class Solution {
  method(prices) {
    let profit = 0;

    let l = 0;
    for (let r = 1; r < prices.length; r++) {
      // Set left to right if right is less (no profit)
      if (prices[r] < prices[l]) l = r;
      profit = Math.max(profit, prices[r] - prices[l]);
    }

    return profit;
  }

  reference(prices) {
    let [left, right, max] = [0, 1, 0];

    while (right < prices.length) {
      const canSlide = prices[right] <= prices[left];
      if (canSlide) left = right;

      const window = prices[right] - prices[left];

      max = Math.max(max, window);
      right++;
    }

    return max;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [7, 1, 5, 3, 6, 4],
  [7, 6, 4, 3, 1],
];
test.quantify(testCases);
